xarray/core/variable.py

```diff
@@ -1212,34 +1212,79 @@ def __setitem__(self, key, value):
         raise TypeError('%s values cannot be modified' % type(self).__name__)
 
     @classmethod
+    def _concat_numpy(cls, variables, positions=None):
+        """
+        Concatenates variables. Works for variables whose dtype is
+        different from numpy.object. If variables' dtype is numpy.object
+        it throws TypeError and "concat" function will use
+        concat_pandas function
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+
+        indexes = [v._data for v in variables]
+
+        if not indexes:
+            data = []
+        else:
+            data = np.concatenate((indexes))
+
+            if positions is not None:
+                indices = nputils.inverse_permutation(
+                    np.concatenate(positions))
+                data = data.take(indices)
+
+        return data
+
+    @classmethod
+    def _concat_pandas(cls, variables, positions=None):
+        """
+        Concatenates variables. This is generic function that
+        handles all cases for which concat_numpy does not work
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+        indexes = [v._data.array for v in variables]
+
+        if not indexes:
+            data = []
+        else:
+
+            data = indexes[0].append(indexes[1:])
+
+            if positions is not None:
+                indices = nputils.inverse_permutation(
+                    np.concatenate(positions))
+                data = data.take(indices)
+
+        return data
+
+    @classmethod
     def concat(cls, variables, dim='concat_dim', positions=None,
                shortcut=False):
-        """Specialized version of Variable.concat for IndexVariable objects.
-
+        """Specialized version of Variable.concat for
+        IndexVariable objects.
         This exists because we want to avoid converting Index objects to NumPy
         arrays, if possible.
         """
         if not isinstance(dim, basestring):
             dim, = dim.dims
 
         variables = list(variables)
+
         first_var = variables[0]
 
         if any(not isinstance(v, cls) for v in variables):
             raise TypeError('IndexVariable.concat requires that all input '
                             'variables be IndexVariable objects')
 
-        indexes = [v._data.array for v in variables]
-
-        if not indexes:
-            data = []
+        # GH1434
+        # Fixes bug: "xr.concat loses coordinate dtype
+        # information with recarrays in 0.9"
+        if np.object in map(lambda var: var.dtype, variables):
```

Comment:
Can you use a comprehension here instead, e.g., `any(var.dtype == np.object for var in variables)`?